src/usr.sbin/vnconfig/vnconfig.8
.\"	$MirOS: src/usr.sbin/vnconfig/vnconfig.8,v 1.8 2014/03/29 11:36:06 tg Exp $
.\"	$OpenBSD: vnconfig.8,v 1.21 2004/09/10 15:37:53 pedro Exp $
.\"
.\" Copyright (c) 2006, 2009, 2014 Thorsten Glaser
.\" Copyright (c) 1993 University of Utah.
.\" Copyright (c) 1980, 1989, 1991, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" This code is derived from software contributed to Berkeley by
.\" the Systems Programming Group of the University of Utah Computer
.\" Science Department.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     @(#)vnconfig.8	8.1 (Berkeley) 6/5/93
.\"
.Dd $Mdocdate: March 29 2014 $
.Dt VNCONFIG 8
.Os
.Sh NAME
.Nm vnconfig
.Nd "configure vnode disks for file swapping or pseudo filesystems"
.Sh SYNOPSIS
.Nm
.Op Fl c
.Op Fl vkr
.Op Fl f Ar keyfile
.Ar rawdev
.Ar regular_file
.Nm
.Fl u
.Op Fl v
.Ar rawdev
.Nm
.Fl l
.Op Ar rawdev
.Nm
.Fl K Ar algorithm
.Fl f Ar keyfile
.Sh DESCRIPTION
The
.Nm
command configures vnode pseudo disk devices.
It will associate (or disassociate) the special file
.Ar rawdev
with the regular file
.Ar regular_file
allowing the latter to be accessed as though it were a disk.
Hence a regular file within the filesystem can be used for swapping
or can contain a filesystem that is mounted in the name space.
Both traditional devices,
.Pa vnd ,
and the cache-coherent devices,
.Pa svnd ,
are configured through
.Nm vnconfig .
.Pp
If you read the regular file while using
.Pa svnd ,
you need to mount the
.Pa svnd
with the mount option
.Ar sync
to ensure consistency in the buffer cache.
For normal operation this is not necessary.
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl c
Configures the device.
If successful, references to
.Ar rawdev
will access the contents of
.Ar regular_file .
This is the default operation.
.It Fl f Ar keyfile
Instead of an encryption key read from stdin, an ASN.1 octet string
will be decoded from the PEM format
.Ar keyfile .
This option implies
.Fl k .
.It Fl k
Associate an encryption key with the device.
All data will be encrypted using the Blowfish cipher before it is
written to the disk.
Encryption only works with
.Pa svnd ,
not with
.Pa vnd .
.It Fl K Ar algorithm
Generate a PEM format keyfile containing an encrypted ASN.1
octet string which can be used as keyfile for further invocations of
.Nm
.Fl kf .
This is incompatible with the
.Fl K Ar rounds / Fl S Ar saltfile
method invented in
.Ox 4.2 .
.Pp
If the
.Fl k
option is also used, the system will convert an old (plain password)
encryption key to a keyfile by first asking (twice) for the plaintext
password, then (twice) for the PEM passphrase with which to protect
the keyfile (consisting of an encoded version of the old plaintext
password).
This can be used to migrate old encrypted volumes using the
.Ox
blowfish (bf-cbc) format to the new keyfile-based scheme, which
supports changing the user (PEM) passphrase without needing to
re-encrypt the actual data.
.Pp
If the
.Fl k
option is not used,
.Nm
will create a keyfile consisting of 448 random bits (for use with
the new-style bf-cbc algorithm \*(en other algorithms will follow
and use different formats), ask (twice) for a PEM passphrase with
which the newly created keyfile is to be protected, which can be
changed by the user without needing to re-encrypt the actual data,
then save the keyfile just created to disc.
.Pp
The
.Ar algorithm
can be chosen from the default
.Xr openssltool 1
list; see below for an example.
.It Fl l
List the vnd devices and indicate which ones are in use.
If a specific
.Ar rawdev
is given, then only that one will be described.
.It Fl r
If the kernel supports it, don't write to the underlying file.
This is supported since
.Mx 7 .
.It Fl u
Unconfigures the device.
.It Fl v
Print messages to stdout describing actions taken.
.El
.Pp
If no action option is given,
.Fl c
is assumed.
.Sh FILES
.Bl -tag -width /etc/rvnd?? -compact
.It Pa /dev/{,r}{,s}vnd*
.El
.Sh EXAMPLES
Configure the vnode disk
.Pa svnd0
(all of its partitions become usable):
.Pp
.Dl # vnconfig svnd0 /tmp/diskimage
.Pp
Unconfigure the same device:
.Pp
.Dl # vnconfig -u svnd0
.Pp
Show a list of supported
.Em Cipher Types :
.Pp
.Dl $ openssl enc -
.Pp
Create an encrypted image and a keyfile:
.Bd -literal -offset indent
$ dd if=/dev/arandom of=/tmp/foo.ima bs=1048576 count=16
$ vnconfig -K AES-256-CBC -f /tmp/foo.skf
.Ed
.Pp
Use an encrypted image with a keyfile:
.Pp
.Dl # vnconfig -f /tmp/foo.skf svnd0 /tmp/foo.ima
.Sh SEE ALSO
.Xr openssl 1 ,
.Xr openssltool 1 ,
.Xr arc4random 3 ,
.Xr blowfish 3 ,
.Xr vnd 4 ,
.Xr mount 8 ,
.Xr swapon 8 ,
.Xr umount 8
.Sh HISTORY
Support for external keyfiles was added in
.Mx 9 .
.Sh BUGS
This command should really be named
.Nm vndconfig .
.Pp
Currently, only bf-cbc (Blowfish) encryption in an
.Ox
compatible format is available.
Support for AES (including VIA C3 hardware acceleration) is planned.
.Pp
Configuring a block device, such as
.Pa /dev/wd0e ,
as
.Ar regular_file
will result in very bad I/O performance due to the use of
synchronous (reliable) writes, even if the inner filesystem
is mounted async or with softupdates.
Character special devices cannot be used at all.
This is a bug inherited from
.Ox .


src/usr.sbin/vnconfig/vnconfig.c
/**	$MirOS: src/usr.sbin/vnconfig/vnconfig.c,v 1.14 2007/08/08 22:20:17 tg Exp $ */
/*	$OpenBSD: vnconfig.c,v 1.16 2004/09/14 22:35:51 deraadt Exp $	*/
/*
 * Copyright (c) 2006, 2010 Thorsten Glaser
 * Copyright (c) 1993 University of Utah.
 * Copyright (c) 1990, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * This code is derived from software contributed to Berkeley by
 * the Systems Programming Group of the University of Utah Computer
 * Science Department.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 * from: Utah $Hdr: vnconfig.c 1.1 93/12/15$
 *
 *	@(#)vnconfig.c	8.1 (Berkeley) 12/15/93
 */

#include <sys/param.h>
#include <sys/ioctl.h>
#include <sys/mount.h>
#include <sys/stat.h>

#include <dev/vndioctl.h>

#include <err.h>
#include <errno.h>
#include <fcntl.h>
#include <pwd.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <util.h>

#ifndef SMALL
#include <openssl/asn1.h>
#include <openssl/err.h>
#include <openssl/evp.h>
#include <openssl/pem.h>
#endif

__RCSID("$MirOS: src/usr.sbin/vnconfig/vnconfig.c,v 1.14 2007/08/08 22:20:17 tg Exp $");

#define DEFAULT_VND	"vnd0"

#define VND_CONFIG	1
#define VND_UNCONFIG	2
#define VND_GET		3
#define VND_MAKEKEY	4

#define BLF_MAX_BYTES	72

int verbose = 0;

__dead void usage(void);
int config(char *, char *, int, char *, size_t, u_int32_t);
int getinfo(const char *);
#ifndef SMALL
static void extract_key(FILE *, char **, int *);
static int make_key(const char *, FILE *, const char *);
#endif

int
main(int argc, char **argv)
{
	int ch, rv, action = VND_CONFIG, flags = 0, keylen = 0;
	char *key = NULL;
#ifndef SMALL
	char *key2 = NULL, *keyfile = NULL, *algo = NULL;
	FILE *keyfp = NULL;
#endif

	while ((ch = getopt(argc, argv, "cf:K:klruv")) != -1) {
		switch (ch) {
		case 'c':
			action = VND_CONFIG;
			break;
#ifndef SMALL
		case 'f':
			keyfile = optarg;
			break;
		case 'K':
			action = VND_MAKEKEY;
			algo = optarg;
			break;
#endif
		case 'l':
			action = VND_GET;
			break;
		case 'u':
			action = VND_UNCONFIG;
			break;
		case 'v':
			verbose = 1;
			break;
		case 'k':
			key = getpass("Encryption key: ");
			if (key == NULL)
				errx(1, "getpass");
			keylen = strlen(key);
			break;
		case 'r':
			flags |= VNDIOC_OPT_RDONLY;
			break;
		default:
			usage();
			/* NOTREACHED */
		}
	}

	argc -= optind;
	argv += optind;

#ifndef SMALL
	if (keyfile) {
		keyfp = fopen(keyfile, (action == VND_MAKEKEY) ? "wb" : "rb");
		if (keyfp == NULL)
			err(1, "fopen");
		if (action != VND_MAKEKEY)
			extract_key(keyfp, &key, &keylen);
	}

	if (key && action == VND_MAKEKEY) {
		key2 = getpass("repeat the key: ");
		if (memcmp(key, key2, 1+keylen))
			errx(1, "keys not identical");
	}
#endif

	if (action == VND_CONFIG && argc == 2)
		rv = config(argv[0], argv[1], action, key, keylen,
		    flags);
	else if (action == VND_UNCONFIG && argc == 1)
		rv = config(argv[0], NULL, action, key, keylen,
		    flags | VNDIOC_OPT_RDONLY);
	else if (action == VND_GET)
		rv = getinfo(argc ? argv[0] : NULL);
#ifndef SMALL
	else if (action == VND_MAKEKEY)
		rv = make_key(algo, keyfp, key2);
#endif
	else
		usage();

#ifndef SMALL
	if (keyfp)
		fclose(keyfp);
#endif
	return (rv);
}

int
getinfo(const char *vname)
{
	int vd, print_all = 0;
	struct vnd_user vnu;
	char *vnamedup;

	if (vname == NULL) {
		vname = DEFAULT_VND;
		print_all = 1;
	}

	if ((vnamedup = strdup(vname)) == NULL)
		err(1, "strdup");
	vd = opendev(vnamedup, O_RDONLY, OPENDEV_PART, NULL);
	free(vnamedup);
	if (vd < 0)
		err(1, "open: %s", vname);

	vnu.vnu_unit = -1;

query:
	if (ioctl(vd, VNDIOCGET, &vnu) == -1) {
		close(vd);
		return (!(errno == ENXIO && print_all));
	}

	fprintf(stdout, "vnd%d: ", vnu.vnu_unit);

	if (!vnu.vnu_ino)
		fprintf(stdout, "not in use\n");
	else
		fprintf(stdout, "covering %s on %s, inode %d\n", vnu.vnu_file,
		    devname(vnu.vnu_dev, S_IFBLK), vnu.vnu_ino);

	if (print_all) {
		vnu.vnu_unit++;
		goto query;
	}

	close(vd);

	return (0);
}

int
config(char *dev, char *file, int action, char *key, size_t klen,
    u_int32_t flags)
{
	struct vnd_ioctl vndio;
	FILE *f;
	char *rdev;
	int rv = 0;

	if (opendev(dev, (flags & VNDIOC_OPT_RDONLY) ? O_RDONLY : O_RDWR,
	    OPENDEV_PART, &rdev) < 0)
		err(4, "%s", rdev);
	f = fopen(rdev, "rw");
	if (f == NULL) {
		warn("%s", rdev);
		rv = -1;
		goto out;
	}
	vndio.vnd_file = file;
	vndio.vnd_key = (u_char *)key;
	vndio.vnd_keylen = klen;
	vndio.vnd_options = flags;

	/*
	 * Clear (un-configure) the device
	 */
	if (action == VND_UNCONFIG) {
		rv = ioctl(fileno(f), VNDIOCCLR, &vndio);
		if (rv)
			warn("VNDIOCCLR");
		else if (verbose)
			printf("%s: cleared\n", dev);
	}
	/*
	 * Configure the device
	 */
	if (action == VND_CONFIG) {
		rv = ioctl(fileno(f), VNDIOCSET, &vndio);
		if (rv)
			warn("VNDIOCSET");
		else if (verbose)
			printf("%s: %llu bytes on %s\n", dev, vndio.vnd_size,
			    file);
	}

	fclose(f);
	fflush(stdout);
 out:
	if (key)
		memset(key, 0, klen);
	return (rv < 0);
}

__dead void
usage(void)
{
	extern const char *__progname;

	(void)fprintf(stderr,
	    "usage: %s [-c] [-krv] [-f keyfile] rawdev regular-file\n"
	    "       %s -u [-v] rawdev\n"
	    "       %s -K algorithm -f keyfile\n"
	    "       %s -l [rawdev]\n",
	    __progname, __progname, __progname, __progname);
	exit(1);
}

#ifndef SMALL
static int
make_key(const char *algo, FILE *fp, const char *key2)
{
	ASN1_OCTET_STRING *aos;
	const EVP_CIPHER *enc = NULL;
#define KBUF_ELEM	(BLF_MAX_BYTES / 4)
	uint32_t kbuf[KBUF_ELEM];
	int i;
	size_t kbuflen = sizeof (kbuf);

	if (algo == NULL || fp == NULL)
		usage();

	ERR_load_crypto_strings();
	OpenSSL_add_all_ciphers();
	if ((aos = ASN1_OCTET_STRING_new()) == NULL)
		errx(2, "cannot create ASN.1 octet string: %s",
		    ERR_error_string(ERR_get_error(), NULL));
	if ((enc = EVP_get_cipherbyname(algo)) == NULL)
		errx(2, "unknown cipher '%s': %s", algo,
		    ERR_error_string(ERR_get_error(), NULL));

	if (key2) {
		if (strlen(key2) < kbuflen)
			kbuflen = strlen(key2);
		else
			fprintf(stderr, "WARNING: truncating from %zu to"
			    " %zu characters!\n", strlen(key2), kbuflen);
		memcpy(kbuf, key2, kbuflen);
		fprintf(stderr, "WARNING: not using random bits as key!\n");
	} else
		arc4random_buf(kbuf, sizeof(kbuf));

	i = ASN1_STRING_set(aos, kbuf, kbuflen);
	bzero(kbuf, sizeof (kbuf));
	if (!i)
		errx(2, "cannot set ASN.1 octet string: %s",
		    ERR_error_string(ERR_get_error(), NULL));

	i = PEM_write_ASN1_OCTET_STRING(fp, aos, enc, NULL, 0, NULL, NULL);
	ASN1_STRING_free(aos);
	if (!i)
		errx(2, "cannot encode svnd keyfile: %s",
		    ERR_error_string(ERR_get_error(), NULL));
	return (0);
}

static void
extract_key(FILE *fp, char **key, int *klen)
{
	ASN1_OCTET_STRING *aos;
	unsigned char *p;

	if (fp == NULL || key == NULL || klen == NULL)
		usage();

	ERR_load_crypto_strings();
	OpenSSL_add_all_ciphers();
	if ((aos = PEM_read_ASN1_OCTET_STRING(fp, NULL, NULL, NULL)) == NULL) {
		ERR_print_errors_fp(stderr);
		errx(2, "cannot decode svnd keyfile");
	}

	*klen = ASN1_STRING_length(aos);
	if (*klen < 1) {
		ASN1_STRING_free(aos);
		ERR_print_errors_fp(stderr);
		errx(2, "svnd keyfile too small: %d", *klen);
	} else if (*klen > BLF_MAX_BYTES)
		*klen = BLF_MAX_BYTES;

	*key = malloc(*klen);
	if (*key == NULL)
		err(1, "malloc of %d bytes", *klen);

	p = ASN1_STRING_data(aos);
	if (p == NULL) {
		ASN1_STRING_free(aos);
		ERR_print_errors_fp(stderr);
		errx(2, "cannot get ASN.1 string");
	}

	memcpy(*key, p, *klen);
	ASN1_STRING_free(aos);
	EVP_cleanup();
	ERR_free_strings();
}
#endif
